File: ta4j-core/src/test/java/org/ta4j/core/analysis/ReturnsTest.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.analysis;

import org.junit.Test;
import org.ta4j.core.BarSeries;
import org.ta4j.core.BaseTradingRecord;
import org.ta4j.core.Trade;
import org.ta4j.core.TradingRecord;
import org.ta4j.core.criteria.ReturnRepresentation;
import org.ta4j.core.mocks.MockBarSeriesBuilder;
import org.ta4j.core.num.*;

import java.math.MathContext;
import java.math.RoundingMode;

import static org.junit.Assert.assertEquals;
import static org.ta4j.core.TestUtils.assertNumEquals;

public class ReturnsTest {

    private NumFactory numFactory = DoubleNumFactory.getInstance();

    @Test
    public void returnSize() {
        // Test with both LOG and DECIMAL representations
        ReturnRepresentation[] representations = { ReturnRepresentation.LOG, ReturnRepresentation.DECIMAL };
        for (ReturnRepresentation representation : representations) {
            // No return at index 0
            var sampleBarSeries = new MockBarSeriesBuilder().withNumFactory(numFactory)
                    .withData(1d, 2d, 3d, 4d, 5d)
                    .build();
            Returns returns = new Returns(sampleBarSeries, new BaseTradingRecord(), representation);
            assertEquals(4, returns.getSize());
        }
    }

    @Test
    public void singleReturnPositionArith() {
        var sampleBarSeries = new MockBarSeriesBuilder().withNumFactory(numFactory).withData(1d, 2d).build();
        TradingRecord tradingRecord = new BaseTradingRecord(Trade.buyAt(0, sampleBarSeries),
                Trade.sellAt(1, sampleBarSeries));
        Returns return1 = new Returns(sampleBarSeries, tradingRecord, ReturnRepresentation.DECIMAL);
        assertNumEquals(NaN.NaN, return1.getValue(0));
        assertNumEquals(1.0, return1.getValue(1));
    }

    @Test
    public void returnsWithSellAndBuyTrades() {
        var sampleBarSeries = new MockBarSeriesBuilder().withNumFactory(numFactory)
                .withData(2, 1, 3, 5, 6, 3, 20)
                .build();
        TradingRecord tradingRecord = new BaseTradingRecord(Trade.buyAt(0, sampleBarSeries),
                Trade.sellAt(1, sampleBarSeries), Trade.buyAt(3, sampleBarSeries), Trade.sellAt(4, sampleBarSeries),
                Trade.sellAt(5, sampleBarSeries), Trade.buyAt(6, sampleBarSeries));

        Returns strategyReturns = new Returns(sampleBarSeries, tradingRecord, ReturnRepresentation.DECIMAL);

        assertNumEquals(NaN.NaN, strategyReturns.getValue(0));
        assertNumEquals(-0.5, strategyReturns.getValue(1));
        assertNumEquals(0, strategyReturns.getValue(2));
        assertNumEquals(0, strategyReturns.getValue(3));
        assertNumEquals(1d / 5, strategyReturns.getValue(4));
        assertNumEquals(0, strategyReturns.getValue(5));
        assertNumEquals(1 - (20d / 3), strategyReturns.getValue(6));
    }

    @Test
    public void returnsWithGaps() {
        var sampleBarSeries = new MockBarSeriesBuilder().withNumFactory(numFactory)
                .withData(1d, 2d, 3d, 4d, 5d, 6d, 7d, 8d, 9d, 10d, 11d, 12d)
                .build();
        TradingRecord tradingRecord = new BaseTradingRecord(Trade.sellAt(2, sampleBarSeries),
                Trade.buyAt(5, sampleBarSeries), Trade.buyAt(8, sampleBarSeries), Trade.sellAt(10, sampleBarSeries));

        Returns returns = new Returns(sampleBarSeries, tradingRecord, ReturnRepresentation.LOG);

        assertNumEquals(NaN.NaN, returns.getValue(0));
        assertNumEquals(0, returns.getValue(1));
        assertNumEquals(0, returns.getValue(2));
        assertNumEquals(-0.28768207245178085, returns.getValue(3));
        assertNumEquals(-0.22314355131420976, returns.getValue(4));
        assertNumEquals(-0.1823215567939546, returns.getValue(5));
        assertNumEquals(0, returns.getValue(6));
        assertNumEquals(0, returns.getValue(7));
        assertNumEquals(0, returns.getValue(8));
        assertNumEquals(0.10536051565782635, returns.getValue(9));
        assertNumEquals(0.09531017980432493, returns.getValue(10));
        assertNumEquals(0, returns.getValue(11));

    }

    @Test
    public void returnsWithNoPositions() {
        var sampleBarSeries = new MockBarSeriesBuilder().withNumFactory(numFactory)
                .withData(3d, 2d, 5d, 4d, 7d, 6d, 7d, 8d, 5d, 6d)
                .build();
        Returns returns = new Returns(sampleBarSeries, new BaseTradingRecord(), ReturnRepresentation.LOG);
        assertNumEquals(NaN.NaN, returns.getValue(0));
        assertNumEquals(0, returns.getValue(4));
        assertNumEquals(0, returns.getValue(7));
        assertNumEquals(0, returns.getValue(9));
    }

    @Test
    public void returnsPrecision() {
        var doubleSeries = new MockBarSeriesBuilder().withNumFactory(numFactory).withData(1.2d, 1.1d).build();
        final var highPrecisionContext = new MathContext(32, RoundingMode.HALF_UP);
        final var precisionFactory = DecimalNumFactory.getInstance(highPrecisionContext);
        BarSeries precisionSeries = new MockBarSeriesBuilder().withNumFactory(precisionFactory)
                .withData(1.2d, 1.1d)
                .build();

        TradingRecord fullRecordDouble = new BaseTradingRecord();
        fullRecordDouble.enter(doubleSeries.getBeginIndex(), doubleSeries.getBar(0).getClosePrice(),
                doubleSeries.numFactory().one());
        fullRecordDouble.exit(doubleSeries.getEndIndex(), doubleSeries.getBar(1).getClosePrice(),
                doubleSeries.numFactory().one());

        TradingRecord fullRecordPrecision = new BaseTradingRecord();
        fullRecordPrecision.enter(precisionSeries.getBeginIndex(), precisionSeries.getBar(0).getClosePrice(),
                precisionSeries.numFactory().one());
        fullRecordPrecision.exit(precisionSeries.getEndIndex(), precisionSeries.getBar(1).getClosePrice(),
                precisionSeries.numFactory().one());

        // Return calculation DoubleNum vs PrecisionNum
        Num arithDouble = new Returns(doubleSeries, fullRecordDouble, ReturnRepresentation.DECIMAL).getValue(1);
        Num arithPrecision = new Returns(precisionSeries, fullRecordPrecision, ReturnRepresentation.DECIMAL)
                .getValue(1);
        Num logDouble = new Returns(doubleSeries, fullRecordDouble, ReturnRepresentation.LOG).getValue(1);
        Num logPrecision = new Returns(precisionSeries, fullRecordPrecision, ReturnRepresentation.LOG).getValue(1);

        assertNumEquals(DoubleNum.valueOf(-0.08333333333333326), arithDouble);
        final var expectedArithmetic = DecimalNum.valueOf("1.1", highPrecisionContext)
                .dividedBy(DecimalNum.valueOf("1.2", highPrecisionContext))
                .minus(DecimalNum.valueOf(1, highPrecisionContext));
        assertNumEquals(arithPrecision, expectedArithmetic);

        assertNumEquals(DoubleNum.valueOf(-0.08701137698962969), logDouble);
        assertNumEquals(DecimalNum.valueOf("-0.087011376989629766167765901873746", highPrecisionContext), logPrecision);
    }
}

File: ta4j-core/src/test/kotlin/org/ta4j/core/analysis/ReturnsTest.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.analysis

import org.junit.Assert
import org.junit.Test
import org.ta4j.core.*
import org.ta4j.core.Trade.Companion.buyAt
import org.ta4j.core.Trade.Companion.sellAt
import org.ta4j.core.indicators.AbstractIndicatorTest
import org.ta4j.core.mocks.MockBarSeries
import org.ta4j.core.num.DecimalNum
import org.ta4j.core.num.DecimalNum.Companion.valueOf
import org.ta4j.core.num.DoubleNum
import org.ta4j.core.num.NaN
import org.ta4j.core.num.Num
import java.util.function.Function

class ReturnsTest(numFunction: Function<Number?, Num>?) : AbstractIndicatorTest<Indicator<Num>?, Num>(
    Function { obj: Number? -> DoubleNum.Companion.valueOf(obj) }) {
    @Test
    fun returnSize() {
        for (type in Returns.ReturnType.values()) {
            // No return at index 0
            val sampleBarSeries: BarSeries = MockBarSeries(numFunction, 1.0, 2.0, 3.0, 4.0, 5.0)
            val returns = Returns(sampleBarSeries, BaseTradingRecord(), type)
            Assert.assertEquals(4, returns.size.toLong())
        }
    }

    @Test
    fun singleReturnPositionArith() {
        val sampleBarSeries: BarSeries = MockBarSeries(numFunction, 1.0, 2.0)
        val tradingRecord: TradingRecord = BaseTradingRecord(
            buyAt(0, sampleBarSeries),
            sellAt(1, sampleBarSeries)
        )
        val return1 = Returns(sampleBarSeries, tradingRecord, Returns.ReturnType.ARITHMETIC)
        TestUtils.assertNumEquals(NaN.NaN, return1[0])
        TestUtils.assertNumEquals(1.0, return1.getValue(1))
    }

    @Test
    fun returnsWithSellAndBuyTrades() {
        val sampleBarSeries: BarSeries = MockBarSeries(numFunction, 2.0, 1.0, 3.0, 5.0, 6.0, 3.0, 20.0)
        val tradingRecord: TradingRecord = BaseTradingRecord(
            buyAt(0, sampleBarSeries),
            sellAt(1, sampleBarSeries), buyAt(3, sampleBarSeries), sellAt(4, sampleBarSeries),
            sellAt(5, sampleBarSeries), buyAt(6, sampleBarSeries)
        )
        val strategyReturns = Returns(sampleBarSeries, tradingRecord, Returns.ReturnType.ARITHMETIC)
        TestUtils.assertNumEquals(NaN.NaN, strategyReturns[0])
        TestUtils.assertNumEquals(-0.5, strategyReturns.getValue(1))
        TestUtils.assertNumEquals(0, strategyReturns.getValue(2))
        TestUtils.assertNumEquals(0, strategyReturns.getValue(3))
        TestUtils.assertNumEquals(1.0 / 5, strategyReturns.getValue(4))
        TestUtils.assertNumEquals(0, strategyReturns.getValue(5))
        TestUtils.assertNumEquals(1 - 20.0 / 3, strategyReturns.getValue(6))
    }

    @Test
    fun returnsWithGaps() {
        val sampleBarSeries: BarSeries =
            MockBarSeries(numFunction, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0)
        val tradingRecord: TradingRecord = BaseTradingRecord(
            sellAt(2, sampleBarSeries),
            buyAt(5, sampleBarSeries), buyAt(8, sampleBarSeries), sellAt(10, sampleBarSeries)
        )
        val returns = Returns(sampleBarSeries, tradingRecord, Returns.ReturnType.LOG)
        TestUtils.assertNumEquals(NaN.NaN, returns[0])
        TestUtils.assertNumEquals(0, returns.getValue(1))
        TestUtils.assertNumEquals(0, returns.getValue(2))
        TestUtils.assertNumEquals(-0.28768207245178085, returns.getValue(3))
        TestUtils.assertNumEquals(-0.22314355131420976, returns.getValue(4))
        TestUtils.assertNumEquals(-0.1823215567939546, returns.getValue(5))
        TestUtils.assertNumEquals(0, returns.getValue(6))
        TestUtils.assertNumEquals(0, returns.getValue(7))
        TestUtils.assertNumEquals(0, returns.getValue(8))
        TestUtils.assertNumEquals(0.10536051565782635, returns.getValue(9))
        TestUtils.assertNumEquals(0.09531017980432493, returns.getValue(10))
        TestUtils.assertNumEquals(0, returns.getValue(11))
    }

    @Test
    fun returnsWithNoPositions() {
        val sampleBarSeries: BarSeries = MockBarSeries(numFunction, 3.0, 2.0, 5.0, 4.0, 7.0, 6.0, 7.0, 8.0, 5.0, 6.0)
        val returns = Returns(sampleBarSeries, BaseTradingRecord(), Returns.ReturnType.LOG)
        TestUtils.assertNumEquals(NaN.NaN, returns[0])
        TestUtils.assertNumEquals(0, returns.getValue(4))
        TestUtils.assertNumEquals(0, returns.getValue(7))
        TestUtils.assertNumEquals(0, returns.getValue(9))
    }

    @Test
    fun returnsPrecision() {
        val doubleSeries: BarSeries = MockBarSeries(numFunction, 1.2, 1.1)
        val precisionSeries: BarSeries =
            MockBarSeries({ obj: Number? -> DecimalNum.Companion.valueOf(obj) }, 1.2, 1.1)
        val fullRecordDouble: TradingRecord = BaseTradingRecord()
        fullRecordDouble.enter(
            doubleSeries.beginIndex, doubleSeries.getBar(0).closePrice,
            doubleSeries.numOf(1)
        )
        fullRecordDouble.exit(
            doubleSeries.endIndex, doubleSeries.getBar(1).closePrice,
            doubleSeries.numOf(1)
        )
        val fullRecordPrecision: TradingRecord = BaseTradingRecord()
        fullRecordPrecision.enter(
            precisionSeries.beginIndex, precisionSeries.getBar(0).closePrice,
            precisionSeries.numOf(1)
        )
        fullRecordPrecision.exit(
            precisionSeries.endIndex, precisionSeries.getBar(1).closePrice,
            precisionSeries.numOf(1)
        )

        // Return calculation DoubleNum vs PrecisionNum
        val arithDouble = Returns(doubleSeries, fullRecordDouble, Returns.ReturnType.ARITHMETIC).getValue(1)
        val arithPrecision = Returns(precisionSeries, fullRecordPrecision, Returns.ReturnType.ARITHMETIC)
            .getValue(1)
        val logDouble = Returns(doubleSeries, fullRecordDouble, Returns.ReturnType.LOG).getValue(1)
        val logPrecision = Returns(precisionSeries, fullRecordPrecision, Returns.ReturnType.LOG).getValue(1)
        TestUtils.assertNumEquals(arithDouble, DoubleNum.valueOf(-0.08333333333333326))
        TestUtils.assertNumEquals(
            arithPrecision,
            valueOf(1.1).div(valueOf(1.2)).minus(valueOf(1))
        )
        TestUtils.assertNumEquals(logDouble, DoubleNum.valueOf(-0.08701137698962969))
        TestUtils.assertNumEquals(logPrecision, valueOf("-0.087011376989629766167765901873746"))
    }
}